File: Affine/affine.py
"""
Implementare a algoritmului de criptare Affine.
"""
import sys


def alphabet_generator(low_char=65, up_char=90):
	"""
	Creează o listă de caractere ASCII, între limitele specificate.
	Limitele implicite sunt: 65=A, 90=Z.
	@param low_char valoarea ASCII pentru limita inferioară
	@param up_char valoarea ASCII pentru limita superioară
	@returns lista
	"""
	result = ""
	for i in [chr(i) for i in range(low_char, up_char + 1)]:
		result += i
	return result


def encrypt(text, key):
	"""
	Criptează un text, folosind cheia dată.
	@param text
	@param key
	@returns text criptat
	"""
	result = ""
	a = key[0]
	b = key[1]

	for i in text:
		l = alphabet.find(i)
		result += alphabet[(a * l + b) % m]
	return result


def decrypt(text, key):
	"""
	Decriptează un text, folosind cheia dată.
	@param text
	@param key
	@returns text decriptat
	"""
	result = ""
	a = key[0]
	b = key[1]
	a_inv = invert(a, m)
	s = 0

	for i in text:
		l = alphabet.find(i)
		result += alphabet[(a_inv * (l - b)) % m]
	return result


def invert(a, m):
	"""
	Calculează inversul (simetricul) unui număr a în mulțimea Zm.
	@param a numărul de inversat în Zm
	@param m dimensiunea mulțimii Z
	@returns inversul lui a în Zm
	"""
	s = 0
	for i in range(1, m):
		if (((a % m) * (s % m)) % m == 1):
			return s
		s += 1


def gcd(a, b):
	"""
	Calculează CMMDC a două numere.
	@param a
	@param b
	@returns CMMDC(a, b)
	"""
	a = abs(a)
	b = abs(b)

	while a != b:
		if a > b:
			a -= b
		else:
			b -= a

	return a


#aliasuri pentru scriere rapidă
e = encrypt
d = decrypt
gen = alphabet_generator
i = invert

#switch între setul ASCII extins și cel de 26 de caractere;
# TODO: ar trebui făcută verificarea inputului în funcție de setul ales...
# utilizare: py -i affine.py -e
if "-e" in str(sys.argv):
	low_char = 32
	up_char = 126
	print("Alfabetul selectat este: 32 <space>... 126 ~")
else:
	low_char = 65
	up_char = 90
	print("Alfabetul selectat este: 65 A... 90 Z")

#alfabetul ASCII extins la caracterele printabile
alphabet = alphabet_generator(low_char, up_char)  # hardcodat pentru teste
m = len(alphabet)
print(f"Lungimea alfabetului: m = {m}")

#input de la utilizator
flag = True
while flag:
	plain_text = t = input("[*] Scrie textul care trebuie criptat (ASCII): ")
	print("[*] Scrie cheia:")
	a = b = 1
	while True:
		try:
			a = int(input("[*] a = "))
			# verificarea condiției: gcd(a, m) = 1
			if gcd(a, m) != 1:
				a = int(
				    input(
				        f"[!] a și m = {m} trebuie să fie coprime.\n[!] Scrie o altă valoare pentru "
				        "a: "))
			b = int(input("[*] b = "))
			break
		except:
			print("[!] Trebuie scrise numere întregi!")

	key = k = (a, b)

	#output
	cipher_text = c = encrypt(plain_text, key)
	print(f"[*] Text criptat:   {cipher_text}")
	print(f"[*] Text decriptat: {decrypt(cipher_text, key)}")

	flag = True if input(
	    "\n[?] Continuăm? (d / n): ").lower() == "d" else False

File: Affine/affine.cpp
/**
 * Cifrul afin.
 * - lungimea alfabetului: m (ex. 26)
 * - cheia: K = (a, b)
 * - conditie: a si m coprime (gcd(a, m) = 1)
 * - functia de criptare: E(x) = (ax + b) mod m
 * - functia de decriptare: D(x) = a_inv (x - b) mod m, unde a_inv este
 * simetricul lui a in Zm
 *
 * Specificatii suplimentare:
 * - alfabetul trebuie sa fie mai flexibil (ex. sa poata include setul ASCII
 * printabil: de la 32 = spatiu la 126 = ~); alfabetul are maparea (ASCII ->
 * index): 32 -> 0; 33 -> 1 ...; 126 -> 94
 * N.B.: programul poate prezenta si urme de validare a inputului :)
 * https://github.com/ucv-cs/Securitatea-sistemelor-informatice
 */

#include <iostream>	 // cin, cout, endl
#include <string>	 // getline()
using namespace std;

// limitele alfabetului utilizat, prin raportare la valorile ASCII
// implicit: A-Z, dar poate deveni <space>-~, daca programul ruleaza cu -e
int low_char = 65;	// = A
int up_char = 90;	// = Z
int m;				// lungimea alfabetului

// cheia K = (a, b)
struct K {
	int a;
	int b;
};

/**
 * Reimplementare a operatiei modulo, pentru rezultate ca in Python.
 * @source https://stackoverflow.com/a/1907585
 *
 * @param a
 * @param b
 * @return modulo implementat ca in Python
 */
int mod(int a, int b) {
	return ((a % b) + b) % b;
}

/**
 * Calculeaza CMMDC a doua numere (necesar pentru a testa daca doua
 * numere sunt coprime = prime intre ele: au CMMDC = 1).
 *
 * @param a primul numar
 * @param b al doilea numar
 * @return CMMDC(a, b)
 */
int gcd(int a, int b) {
	// normalizeaza numerele, intrucat: GCD(a, b) = GCD(|a|, |b|)
	a = abs(a);
	b = abs(b);

	while (a != b) {
		if (a > b) {
			a -= b;
		} else {
			b -= a;
		}
	}

	return a;
}

/**
 * Calculeaza inversul (simetricul) unui numar a in multimea Zm.
 *
 * @param a numarul de inversat in Zm
 * @param m dimensiunea multimii Z
 * @return inversul lui a in Zm
 */
int invert(int a, int m) {
	for (int s = 1; s < m; s++) {
		if (mod((mod(a, m) * mod(s, m)), m) == 1) {
			return s;
		}
	}

	return 0;
}

/**
 * Realizeaza criptarea pentru fiecare caracter dupa formula: E(x) = (ax + b)
 * mod m.
 *
 * @param text inputul
 * @param key cheia
 * @return stringul criptat
 */
string encrypt(string text, struct K key) {
	string result;
	for (int i = 0; i < text.length(); i++) {
		result += char(mod(key.a * (text[i] - low_char) + key.b, m) + low_char);
	}

	return result;
}

/**
 * Realizeaza decriptarea pentru fiecare caracter dupa formula: D(x) = a_inv (x
 * - b) mod m.
 *
 * @param text inputul
 * @param key cheia
 * @return stringul decriptat
 */
string decrypt(string text, struct K key) {
	string result;
	int a_inv = invert(key.a, m);
	for (int i = 0; i < text.length(); i++) {
		result += char(mod(a_inv * (text[i] - low_char - key.b), m) + low_char);
	}
	return result;
}

/**
 * Pregateste cin pentru un nou input.
 */
void clear_cin() {
	cin.clear();			 // reseteaza starea bufferului cin
	cin.ignore(1000, '\n');	 // "goleste" bufferul cin
}

/**
 * Executia programului.
 */
int main(int argc, const char* argv[]) {
	// citeste argumentele programului
	for (int i = 1; i < argc; i++) {
		if (string(argv[i]) == "-e") {
			// limitele setului ASCII printabil
			// (implicit sunt A-Z)
			low_char = 32;	// = spatiu
			up_char = 126;	// = ~
		}
	}

	// lungimea alfabetului
	m = 1 + up_char - low_char;

	string plain_text;
	string cipher_text;
	string separator = "\n------------------------\n";
	struct K key;

	// afiseaza configuratia
	cout << "Configuratie\nAlfabet:\n";
	for (int i = 0; i < m; i++) {
		cout << char(i + low_char) << " ";
		if ((i + 1) % 12 == 0) {
			cout << endl;
		}
	}
	cout << "\nLungime: m = " << m << separator;

	bool flag = true;
	while (flag) {
		// obtine de la utilizator textul in clar
		cout << "[*] Scrie textul care trebuie criptat (ASCII): ";
		getline(cin, plain_text);

		// obtine cheia si verifica indeplinirea conditiei: a si m coprime
		//  <=> (gcd(a, m) = 1); repeta solicitarea pana cand conditia se
		//  indeplineste
		cout << "[*] Scrie cheia (numerele a si b separate prin spatiu): ";
		while (!(cin >> key.a >> key.b) || key.a == 0) {
			cout << "[!] Trebuie scrise doua numere intregi nenule: ";
			clear_cin();
		}

		while (gcd(key.a, m) != 1) {
			cout << "[!] a si m = " << m
				 << " trebuie sa fie coprime.\n[!] Scrie o alta valoare pentru "
					"a: ";
			while (!(cin >> key.a) || key.a == 0) {
				clear_cin();
				cout << "[!] Trebuie scris un numar intreg nenul: ";
			}
			clear_cin();
		}

		// afiseaza textul criptat si pe cel decriptat (ca verificare)
		cipher_text = encrypt(plain_text, key);
		cout << "\n[*] Text criptat:   " << cipher_text << endl;
		cout << "[*] Text decriptat: " << decrypt(cipher_text, key) << endl;

		// controlul repetarii
		cout << "\n[?] Continuam? (d / n): ";
		char answer;
		cin >> answer;
		clear_cin();
		flag = (answer == 'd') ? true : false;
		cout << separator;
	}
}